### PR TITLE
[Automations] Add filtering by tag

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

Adds filtering by tag: https://linear.app/dagster-labs/issue/FE-587/filter-automation-and-job-page-results-by-tag#comment-5bbf1b23

## How I Tested These Changes

Screenshot 2024-10-10 at 10 06 16 AM
Screenshot 2024-10-10 at 10 06 07 AM

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext/WorkspaceQueries.tsx [hunk 1]
```diff
@@ -106,11 +110,19 @@ export const LOCATION_WORKSPACE_QUERY = gql`
...
+    }
   }
 
   fragment WorkspaceSensor on Sensor {
     id
     name
+    tags {
+      key
+      value
+    }
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext/WorkspaceQueries.tsx:125]: ahh this is exciting

File: js_modules/dagster-ui/packages/ui-core/src/ui/Filters/useDefinitionTagFilter.tsx [hunk 2]
```diff
@@ -39,27 +42,45 @@ export const useAssetTagFilter = ({
...
+  (tag: Tag) => {
     return tag;
   },
   (tag) => [tag.key, tag.value].join('|@-@|'),
 );
 
-export function useAssetTagsForAssets(
-  assets: {definition?: {tags?: DefinitionTag[] | null} | null}[],
-): DefinitionTag[] {
+export function useTagsForAssets(assets: {definition?: {tags?: Tag[] | null} | null}[]) {
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/ui/Filters/useDefinitionTagFilter.tsx:64]: Only partially related to this line, but I've noticed that sometimes these new resolvers should be non-null and we just forget to use the non_null_list util in python. I'm not sure under what circumstance a definition would be present but tags would be null. Maybe this typing was an error and we can fix it?

File: js_modules/dagster-ui/packages/ui-core/src/graph/KindTags.tsx [hunk 3]
```diff
@@ -13,13 +13,16 @@ export const STORAGE_KIND_TAG = 'dagster/storage_kind';
 export const KIND_TAG_PREFIX = `dagster/kind/`;
 
 // Older code servers may be using the legacy compute kind tag, so we need to check for both
-export const isCanonicalComputeKindTag = (tag: DefinitionTag) =>
+export const isCanonicalComputeKindTag = (tag: Omit<DefinitionTag, '__typename'>) =>
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/graph/KindTags.tsx:16]: This seems good but these could just be `{key: string}` which might be more correct becasue they're not always definition tags? (Or maybe they are all called that now)

File: js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx [hunk 4]
```diff
@@ -110,13 +109,38 @@ export const MergedAutomationRoot = () => {
...
   }, [cachedData, visibleRepos]);
 
+  const tagsFilter = useDefinitionTagFilterWithManagedState({
+    allTags: useTagsForObjects(
+      repoBuckets,
+      useCallback((repoBucket) => {
+        return [
+          ...repoBucket.schedules.flatMap((schedule) => schedule.tags),
+          ...repoBucket.sensors.flatMap((sensor) => sensor.tags),
+        ];
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx:119]: Is it ok for this to return 1000x copies of the same tag? It seems like that could happen here 👀

File: js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx [hunk 5]
```diff
@@ -110,13 +109,38 @@ export const MergedAutomationRoot = () => {
...
 
   const filteredBuckets = useMemo(() => {
     return repoBuckets.map(({sensors, schedules, ...rest}) => {
       return {
         ...rest,
-        sensors: sensors.filter(({sensorState}) => {
+        sensors: sensors.filter(({sensorState, tags}) => {
+          if (
+            tagFilterState.size &&
+            !doesFilterArrayMatchValueArray(Array.from(tagFilterState), tags)
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx:140]: Just a drive-by nit but it's not clear what "match" means here without clicking into the definition. Maybe "doesValueContainAnyFilter" or "doesValueEqualFilter"?

**Review by bengotow — APPROVED** [review]
This all looks great! Left a few inline comments but nothing blocking 🙌

File: js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx (hunk 4 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/automation/MergedAutomationRoot.tsx:119]: `useTagsForObjects` dedupes them
